syntax = "proto3";

/*
  Object meant to ship a file across the wire.
  To preserve data, the file info is only sent once. Followed by chunks
 */
message FileTransfer {
  oneof data {
    FileInfo fileInfo = 1;
    Chunk buffer = 2;
  }
}

// Represents a segment of data
message Chunk {
  // Segment of file being transferred
  bytes buffer = 1;
}

// Contains basic information about a file that's about to be transmitted
message FileInfo {
  // Total length of file being transmitted
  int64 totalLength = 1;

  // Name of file being transmitted
  string filename = 2;
}

/*
  Response to a file being transmitted
*/
message TransferResponse {
  // Total length received
  int64 length = 1;

  // Status code of transaction
  int32 statusCode = 2;

  // Message (if any) about the transaction. Reserved for when an error occurs
  optional string message = 3;
}

/*
  As a server, enable clients to transmit a file
*/
service TransferService {

  // Transfer a file to server
  rpc Transfer(stream FileTransfer) returns (TransferResponse){};
}